File: ex/ieeextreme10/pali.go
package main

import "fmt"

func main() {
    var T int
    var N int
    var name string

    fmt.Scanf("%d\n", &T)
    for t := 0; t < T; t++ {
        fmt.Scanf("%d %s\n", &N, &name)
        stuff := calculate(N, name)
        fmt.Printf("%d\n", len(stuff) % 1000000007)
    }
}

func calculate(n int, name string) map[string]bool {
    stuff := make(map[string]bool, 0)
    changes := make(map[int]bool, 0)
    stuff = npali(n, name, stuff, changes)
    return stuff
}

func npali(n int, name string, stuff map[string]bool, changes map[int]bool) map[string]bool {
    limit := len(name) - 1

    if n == 0 {
        return stuff
    }

    for i := 0; i <= limit; i++ {
        newChanges := copyMap(changes)
        newChanges[i] = true
        for c := 'a'; c <= 'z'; c++ {
            _, here := changes[i]
            if (rune(name[i]) != c) && (here == false) {
                newWord := build(name, i, c)
                if pali(newWord) && newWord != name {
                    stuff[newWord] = true
                }
                stuff = npali(n-1, newWord, stuff, newChanges)
            }
        }
    }

    return stuff
}

func build(word string, i int, c rune) string {
    newWord := []rune(word)
    newWord[i] = c
    return string(newWord)
}

func pali(word string) bool {
    fact := true
    half := len(word) / 2
    limit := len(word) - 1

    for i := 0; i < half && fact; i++ {
        if word[i] != word[limit-i] {
            fact = false
        }
    }

    return fact
}

func copyMap(inlet map[int]bool) map[int] bool {
    outlet := make(map[int]bool, 0)

    for key, value := range inlet {
        outlet[key] = value
    }

    return outlet
}

File: src/ieeextreme10/pali.py
def main(args):
    n = int(args[0])
    name = args[1]

    return len(loop(n, name))

def loop(n, name):
    stuff = set()

    for i in range(1, n+1):
        stuff += npali(i, name)

    return stuff

def npali(n, word):
    stuff = set()

    for i in range(0, n):
        new_word = build(word, i)
        if pali(word):
            stuff.add(new_word)
        stuff += npali(0, new_word)

    return stuff


def build(w, i):
    nw = ''
    for j, it in enumerate(w):
        if j == len(w) -1-i:
            nw += w[i]
        else:
            nw += it
    return nw

def pali(word):
    fact = True
    i = 0

    while i <= len(word)/2:
        fact &= word[i] is word[-1-i]
        i += 1

    return fact

if __name__ == '__main__':
    for _ in range(int(input())):
        print(main(input().split(' ')))
